src/main/java/build/buf/protovalidate/CustomOverload.java

```diff
@@ -611,40 +656,583 @@ private static boolean validateURI(String val, boolean checkAbsolute) {
+  boolean address() {
+    return this.addressPart() && this.index == this.str.length();
+  }
+
+  // Parses an IPv4 Address prefix.
+  boolean addressPrefix() {
+    return this.addressPart()
+        && this.take('/')
+        && this.prefixLength()
+        && this.index == this.str.length();
+  }
+
+  // Stores value in `prefixLen`
+  private boolean prefixLength() {
+    int start = this.index;
+
+    while (true) {
+      if (this.index >= this.str.length() || !this.digit()) {
+        break;
+      }
+
+      if (this.index - start > 2) {
+        // max prefix-length is 32 bits, so anything more than 2 digits is invalid
+        return false;
+      }
+    }
+
+    String str = this.str.substring(start, this.index);
+    if (str.length() == 0) {
+      // too short
+      return false;
+    }
+
+    if (str.length() > 1 && str.charAt(0) == '0') {
+      // bad leading 0
+      return false;
+    }
+
     try {
-      str = new IPAddressString(prefix);
-      addr = str.toAddress();
-    } catch (Exception e) {
+      int val = Integer.parseInt(str);
+
+      if (val > 32) {
+        // max 32 bits
+        return false;
+      }
+
+      this.prefixLen = val;
+
+      return true;
+
+    } catch (NumberFormatException nfe) {
+      // Error converting to number
       return false;
     }
-    if (!addr.isPrefixed()) {
+  }
+
+  private boolean addressPart() {
+    int start = this.index;
+
+    if (this.decOctet()
+        && this.take('.')
+        && this.decOctet()
+        && this.take('.')
+        && this.decOctet()
+        && this.take('.')
+        && this.decOctet()) {
+      return true;
+    }
+
+    this.index = start;
+
+    return false;
+  }
+
+  private boolean decOctet() {
+    int start = this.index;
+
+    while (true) {
+      if (this.index >= this.str.length() || !this.digit()) {
+        break;
+      }
+
+      if (this.index - start > 3) {
+        // decimal octet can be three characters at most
+        return false;
+      }
+    }
+
+    String str = this.str.substring(start, this.index);
+    if (str.length() == 0) {
+      // too short
       return false;
     }
-    if (strict) {
-      IPAddress mask = addr.getNetworkMask().withoutPrefixLength();
-      if (!addr.mask(mask).equals(str.getHostAddress())) {
+
+    if (str.length() > 1 && str.charAt(0) == '0') {
+      // bad leading 0
+      return false;
+    }
+
+    try {
+      int val = Integer.parseInt(str);
+
+      if (val > 255) {
         return false;
       }
+
+      this.octets.add((short) val);
+
+      return true;
+
+    } catch (NumberFormatException nfe) {
+      // Error converting to number
+      return false;
     }
-    if (ver == 0L) {
+  }
+
+  /**
+   * Reports whether the current position is a digit.
+   *
+   * <p>Method parses the rule:
+   *
+   * <p>DIGIT = %x30-39 ; 0-9
+   */
+  private boolean digit() {
+    char c = this.str.charAt(this.index);
+    if ('0' <= c && c <= '9') {
```

Comment:
We've got lots of places in here that could benefit from a `static boolean isDigit(char c)` helper to simplify code.